crates/typed-store/src/rocks/mod.rs

```diff
@@ -133,110 +131,23 @@ macro_rules! reopen {
-        $transaction:expr,
-        $max_retries:expr // should be an Option<int type>, None for unlimited
-        $(,)?
-
-    ) => {{
-        use std::time::Duration;
-
-        use rand::{
-            distributions::{Distribution, Uniform},
-            rngs::ThreadRng,
-        };
-        use tracing::{error, info};
-
-        let mut retries = 0;
-        let max_retries = $max_retries;
-        loop {
-            let status = $transaction;
-            match status {
-                Err(TypedStoreError::RetryableTransaction) => {
-                    retries += 1;
-                    // Randomized delay to help racing transactions get out of each other's way.
-                    let delay = {
-                        let mut rng = ThreadRng::default();
-                        Duration::from_millis(Uniform::new(0, 50).sample(&mut rng))
-                    };
-                    if let Some(max_retries) = max_retries {
-                        if retries > max_retries {
-                            error!(?max_retries, "max retries exceeded");
-                            break status;
-                        }
-                    }
-                    if retries > 10 {
-                        // TODO: monitoring needed?
-                        error!(?delay, ?retries, "excessive transaction retries...");
-                    } else {
-                        info!(
-                            ?delay,
-                            ?retries,
-                            "transaction write conflict detected, sleeping"
-                        );
-                    }
-                    std::thread::sleep(delay);
-                }
-                _ => break status,
-            }
-        }
-    }};
-}
-
-#[macro_export]
-macro_rules! retry_transaction_forever {
-    ($transaction:expr) => {
-        $crate::retry_transaction!($transaction, None)
-    };
-}
-
 #[derive(Debug)]
-pub struct DBWithThreadModeWrapper {
+pub struct RocksDB {
     pub underlying: rocksdb::DBWithThreadMode<MultiThreaded>,
     pub metric_conf: MetricConf,
     pub db_path: PathBuf,
 }
 
-impl DBWithThreadModeWrapper {
-    fn new(
-        underlying: rocksdb::DBWithThreadMode<MultiThreaded>,
-        metric_conf: MetricConf,
-        db_path: PathBuf,
-    ) -> Self {
-        DBMetrics::get().increment_num_active_dbs(&metric_conf.db_name);
-        Self {
-            underlying,
-            metric_conf,
-            db_path,
-        }
-    }
-}
-
-impl Drop for DBWithThreadModeWrapper {
+impl Drop for RocksDB {
     fn drop(&mut self) {
+        self.underlying.cancel_all_background_work(/* wait */ true);
         DBMetrics::get().decrement_num_active_dbs(&self.metric_conf.db_name);
     }
 }
 
-#[derive(Debug)]
-pub struct OptimisticTransactionDBWrapper {
-    pub underlying: rocksdb::OptimisticTransactionDB<MultiThreaded>,
-    pub metric_conf: MetricConf,
-    pub db_path: PathBuf,
-}
-
-impl OptimisticTransactionDBWrapper {
+impl RocksDB {
```

Comment:
Though it's from upstream but I guess we can have one single `impl RocksDB {` block which might be cleaner.